### PR TITLE
Add make target to generate addon config CRs

### LINKED ISSUE #1770: TKR release tool integration for Config CRs
Downstream release tool is creating TKr package that contains the following CRs

1. TanzuKubernetesRelease
2. ClusterBootstrapTemplate
3. OSImage
4. Various AddonConfig CRs

While the tool has enough knowledge to build TKR, CBT, OSImage the config CRs are pretty prescriptive and release tool has no knowledge to generate them. 

This issue is to provide automation to generate config CRs for all addons that support provider model

1. AntreaConfig
2. CalicoConfig
3. KappControllerConfig
----seperate issue for below when the associated PRs are merged
4. VsphereCPIConfig
5. VsphereCSIConfig

Create make targets in tanzu-framework that takes the following as input
1. tkr version (v1.22.1+vmware.1-tkg.1)
2. infra (aws/azure/vsphere-paravirt/vsphere-nonparavirt) 

output
1. AddonConfig CRs

Requirements
1. Config CR is per infra, VsphereCPIConfig and VsphereCSIConfig differ between paravirt and non-paravirt and should not be included for aws, azure
2. Config CR could be separate per TKR

### PR DESCRIPTION
**Pull request by blc1996**
Signed-off-by: Lucheng Bao <luchengb@vmware.com>

### What this PR does / why we need it
Add make target to generate addon config CRs

e.g.
`$ make generate-package-config apiGroup=cni.tanzu.vmware.com kind=AntreaConfig version=v1alpha1 tkr=v1.22.5_vmware.1-tkg.1`

### Which issue(s) this PR fixes

Fixes #1770

### Describe testing done for PR
Locally tested the make target that generates the correct content

### Release note

```release-note
package-based-lcm: Add make target to generate addon config CRs
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by blc1996** [comment]
Hi @ZihanJiang96 @vijaykatam , please let me know if anything needs to be changed. Thanks!

**Comment by ZihanJiang96** [comment]
1. the tkr version parameter wil have value`v1.21.2---vmware.1-tkg.1` instead of `v1.21.2_vmware.1-tkg.1`, please update the Makefile comment?

2. please double confirm with @imikushin about the `.metadata.namespace`, according to my last conversation with him, the CBT we ship will leave this field as empty, not sure about this addons config CR

**Review by maralavi — APPROVED** [review]

File: addons/config/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+Typical templating needed would be `Name` and `Namespace`
+
+``` yaml
+metadata:
+  name: #@ data.values.TKR_VERSION
+  namespace: #@ data.values.GLOBAL_NAMESPACE
+```
+
+Other values should be included if the value should be different than the default value set for the config CR. You can refer to the kubebuilder definations in `apis/` to find the default values.
```
Comment (maralavi) [inline addons/config/README.md:33]: Here you've got linter comment: ```definations``` ---> ```definitions```

File: addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#@ load("@ytt:data", "data")
+---
+apiVersion: run.tanzu.vmware.com/v1alpha3
+kind: KappControllerConfig
+metadata:
+  name: #@ data.values.TKR_VERSION
+  namespace: #@ data.values.GLOBAL_NAMESPACE
+spec:
+  namespace: test-ns
```
Comment (vijaykatam) [inline addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml:9]: Should be `data.values.GLOBAL_NAMESPACE`

File: addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        - key: CriticalAddonsOnly
+          operator: Exists
+        - effect: NoSchedule
+          key: node-role.kubernetes.io/master
+        - effect: NoSchedule
+          key: node.kubernetes.io/not-ready
+        - effect: NoSchedule
+          key: node.cloudprovider.kubernetes.io/uninitialized
+          value: "true"
+
```
Comment (vijaykatam) [inline addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml:24]: Could we default other vars as well?
```
      concurrency: 4
      hostNetwork: true
      metricsBindAddress: "0"
      priorityClassName: system-cluster-critical
```

File: addons/config/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Addon Config CR templates
+
+This folder stores YTT template for generating default addon config CRs used by ClusterBootstrapTemplate. It will be invoked by the make targed at the root of tanzu-framework.
+
+The make target needs 4 inputs, which are GVK of the config CR, and the current TKR version.
+
+``` bash
+make generate-package-config apiGroup=cni.tanzu.vmware.com kind=AntreaConfig version=v1alpha1 tkr=v1.23.3---vmware.1-tkg.1
```
Comment (vijaykatam) [inline addons/config/README.md:8]: Could we document namespace param as well?

**Review by vijaykatam — COMMENTED** [review]
Lets add basic sanity test to verify generated manifests.

File: addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml (hunk 3 above)
Comment (blc1996) [reply addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml:24]: These have the same value as the default of config API. Do we need to reiterate on them?

File: addons/config/README.md (hunk 1 above)
Comment (blc1996) [reply addons/config/README.md]: Fixed

File: addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml (hunk 2 above)
Comment (blc1996) [reply addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml]: Fixed

File: addons/config/README.md (hunk 4 above)
Comment (blc1996) [reply addons/config/README.md]: Documented

**Comment by blc1996** [comment]
> Lets add basic sanity test to verify generated manifests.

Basic sanity template tests are added. Successful sanity tests will be pre-requisite for running YTT templating.

File: addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml (hunk 3 above)
Comment (vijaykatam) [reply addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml]: Lets be consistent. If we are defining some attributes, why not all.

Comment (vijaykatam) [reply addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml]: Lets be consistent. If we are defining some attributes, why not all.

File: addons/config/hack/test.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
```
Comment (vijaykatam) [inline addons/config/hack/test.sh:1]: Could we invoke this in the test target of addons/Makefile?

[the remark this answers is not in the document]
Comment (blc1996) [reply addons/config/hack/test.sh:1]: Added tests to addons make target

File: addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml (hunk 3 above)
Comment (blc1996) [reply addons/config/run.tanzu.vmware.com/v1alpha3/kappcontrollerconfig.yaml]: The attributes I defined have different default values, that's the reason for calling them out explicitly.

**Review by vijaykatam — APPROVED** [review]

**Pull request merged**